File: esp32/7474.c
// Simulate D flip-flop states
bool FF_A = 0;
bool FF_B = 0;

void setup() {
  pinMode(2, OUTPUT); // LED indicator for Y
}

void loop() {
  // Simulated input C (you can alternate or randomize it)
  for (int i = 0; i < 8; i++) {
    bool A = (i & 0x4) >> 2; // D input for FF_A
    bool B = (i & 0x2) >> 1; // D input for FF_B
    bool C = (i & 0x1);

    // Update FFs to new state (simulate clock edge)
    FF_A = A;
    FF_B = B;

    // Calculate Y from flip-flop outputs
    bool Y = (FF_A && FF_B && !C) || (FF_A && !FF_B && C);

    digitalWrite(2, Y); // LED ON if Y=1

    delay(500); // Wait before next input combination
  }
}

File: pico/7474.py
from machine import Pin
import time

# Output pins representing Q of both flip-flops
ff_A = Pin(0, Pin.OUT)
ff_B = Pin(1, Pin.OUT)

# Simulate changing D inputs (A, B)
states = [(0,0), (0,1), (1,0), (1,1)]

while True:
    for D_A, D_B in states:
        # On clock tick, update Q outputs
        ff_A.value(D_A)
        ff_B.value(D_B)
        
        # Optional: Print Q states for debugging
        print("Q_A:", D_A, "Q_B:", D_B)
        
        time.sleep(1)
